File: R/gilded_rose.R
source('item.R')

update_quality <- function(items) {
  lapply(items, 
    function(item) {
      
      if (item$name != "Aged Brie" && item$name != "Backstage passes") {
        if (item$quality > 0) {
          if (item$name != "Diamond") {
            item$quality <- item$quality - 1
          }
        }
      } else {
        if (item$quality < 50) {
          item$quality <- item$quality + 1
          if (item$name == "Backstage passes") {
            if (item$sell_in < 11) {
              if (item$quality < 50) {
                item$quality = item$quality + 1
              }
            }
            if (item$sell_in < 6) {
              if (item$quality < 50) {
                item$quality = item$quality + 1
              }
            }
          }
        }
      }
      
      if (item$name != "Diamond") {
        item$sell_in <- item$sell_in - 1
      }
      
      if (item$sell_in < 0) {
        if (item$name != "Aged Brie") {
          if (item$name != "Backstage passes") {
            if (item$quality > 0) {
              if (item$name != "Diamond") {
                item$quality <- item$quality - 1
              }
            }
          } else {
            item$quality <- item$quality - item$quality
          }
        } else {
          if (item$quality < 50) {
            item$quality <- item$quality + 1
          }
        }
      }
      
      item
    } 
  ) 
}

File: elixir/lib/gilded_rose.ex
defmodule GildedRose do
  # Example
  # update_quality([%Item{name: "Backstage passes", sell_in: 9, quality: 1}])
  # => [%Item{name: "Backstage passes", sell_in: 8, quality: 3}]

  def update_quality(items) do
    Enum.map(items, &update_item/1)
  end

  def update_item(item) do
    item = cond do
      item.name != "Aged Brie" && item.name != "Backstage passes" ->
        if item.quality > 0 do
          if item.name != "Diamond" do
            %{item | quality: item.quality - 1}
          else
            item
          end
        else
          item
        end
      true ->
        cond do
          item.quality < 50 ->
            item = %{item | quality: item.quality + 1}
            cond do
              item.name == "Backstage passes" ->
                item = cond do
                  item.sell_in < 11 ->
                    cond do
                      item.quality < 50 ->
                        %{item | quality: item.quality + 1}
                      true -> item
                    end
                  true -> item
                end
                cond do
                  item.sell_in < 6 ->
                    cond do
                      item.quality < 50 ->
                        %{item | quality: item.quality + 1}
                      true -> item
                    end
                  true -> item
                end
              true -> item
            end
          true -> item
        end
    end
    item = cond do
      item.name != "Diamond" ->
        %{item | sell_in: item.sell_in - 1}
      true -> item
    end
    cond do
      item.sell_in < 0 ->
        cond do
          item.name != "Aged Brie" ->
            cond do
              item.name != "Backstage passes" ->
                cond do
                  item.quality > 0 ->
                    cond do
                      item.name != "Diamond" ->
                        %{item | quality: item.quality - 1}
                      true -> item
                    end
                  true -> item
                end
              true -> %{item | quality: item.quality - item.quality}
            end
          true ->
            cond do
              item.quality < 50 ->
                %{item | quality: item.quality + 1}
              true -> item
            end
        end
      true -> item
    end
  end
end
